pkg/pdp/aggregator/fns/fns.go

```diff
@@ -0,0 +1,147 @@
+package fns
+
+import (
+	"cmp"
+	"context"
+
+	// for go:embed
+	_ "embed"
+	"fmt"
+	"slices"
+
+	ipldprime "github.com/ipld/go-ipld-prime"
+	"github.com/ipld/go-ipld-prime/schema"
+	"github.com/storacha/go-capabilities/pkg/pdp"
+	"github.com/storacha/go-piece/pkg/piece"
+	"github.com/storacha/go-ucanto/core/invocation/ran"
+	"github.com/storacha/go-ucanto/core/ipld"
+	"github.com/storacha/go-ucanto/core/iterable"
+	"github.com/storacha/go-ucanto/core/receipt"
+	"github.com/storacha/go-ucanto/core/result"
+	"github.com/storacha/go-ucanto/ucan"
+	"github.com/storacha/storage/pkg/pdp/aggregator/aggregate"
+	"github.com/storacha/storage/pkg/pdp/curio"
+)
+
+//go:embed piecessofar.ipldsch
+var piecesSoFarSchema []byte
+
+var piecesSoFarTS *schema.TypeSystem
+
+func init() {
+	ts, err := ipldprime.LoadSchemaBytes(piecesSoFarSchema)
+	if err != nil {
+		panic(fmt.Errorf("loading blob schema: %w", err))
+	}
+	piecesSoFarTS = ts
+}
+
+func PiecesSoFarType() schema.Type {
+	return piecesSoFarTS.TypeByName("PiecesSoFarType")
+}
+
+// PiecesSoFar tracks in progress work building an aggregation
+type PiecesSoFar struct {
```

Comment:
We call this a "buffer" in current filecoin pipeline...IDK maybe worth naming it something along those lines?

Reply:
like that